### PR TITLE
splict gc from restore

### PR DESCRIPTION
**Pull request by OsmondJiang**
make the gc as new command of docfx
`docfx gc` to gc AppData
will update the README.md later

Implement #3082

### CODE REVIEW

File: src/docfx/cli/Program.cs [hunk 1]
```diff
@@ -98,6 +102,11 @@ private static (string command, string docset, CommandLineOptions options) Parse
                 syntax.DefineOption("github-token", ref options.GitHubToken, "The GitHub token used to get contribution information from GitHub API");
                 syntax.DefineOption("locale", ref options.Locale, "The locale of the docset to build");
                 syntax.DefineParameter("docset", ref docset, "Docset directory that contains docfx.yml/docfx.json.");
+
+                // GC command
+                // usage: docfx gc [-d/--days days]
+                syntax.DefineCommand("gc", ref command, "Grabage collect for `AppData` folder");
+                syntax.DefineOption("d|days", ref options.NotAccessedDays, "Collect files not accessed for <d> days");
```
Comment (yufeih) [inline src/docfx/cli/Program.cs]: --retention-days ?

File: src/docfx/cli/Program.cs [hunk 2]
```diff
@@ -59,6 +59,10 @@ internal static async Task<int> Run(string[] args)
                             await Build.Run(docset, options, report);
                             Done(stopwatch.Elapsed, report);
                             break;
+                        case "gc":
+                            await GarbageCollector.Collect(options.NotAccessedDays);
+                            Done(stopwatch.Elapsed, report);
```
Comment (yufeih) [inline src/docfx/cli/Program.cs]: gc result is a bit different from build, restore. It does not have error/warning count.

Can we instead show how may files are removed from download cache, how many git repo checkouts have been removed?

File: src/docfx/gc/GarbageCollector.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace Microsoft.Docs.Build
+{
+    internal static class GarbageCollector
+    {
+        private const int MaxKeepingDays = 15;
+
+        public static async Task Collect(int notAccessedForDays)
+        {
+            notAccessedForDays = notAccessedForDays <= 0 ? MaxKeepingDays : notAccessedForDays;
```
Comment (yufeih) [inline src/docfx/gc/GarbageCollector.cs]: respect command line input, make 15 a default value using `CommandLineOptions { public int RetentionDays = 15; }`

File: src/docfx/gc/GarbageCollector.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                return;
+            }
+
+            var gitWorkTreeRoots = Directory.EnumerateDirectories(AppData.GitRestoreDir, ".git", SearchOption.AllDirectories);
+
+            using (Progress.Start("GC dependency git repositories"))
+            {
+                await ParallelUtility.ForEach(
+                    gitWorkTreeRoots,
+                    async gitWorkTreeRoot =>
```
Comment (yufeih) [inline src/docfx/gc/GarbageCollector.cs]: can this be extracted to a function ?

File: src/docfx/gc/GarbageCollector.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        private static void CollectUrls(int notAccessedForDays)
+        {
+            if (!Directory.Exists(AppData.UrlRestoreDir))
+            {
+                return;
+            }
+
+            var downloadedFiles = Directory.EnumerateFiles(AppData.UrlRestoreDir, "*", SearchOption.AllDirectories).Select(f => PathUtility.NormalizeFile(f)).ToList();
+
+            using (Progress.Start("GC downloaded urls"))
```
Comment (yufeih) [inline src/docfx/gc/GarbageCollector.cs]: 'Cleaning downloaded files' ?

File: src/docfx/gc/GarbageCollector.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        private static async Task CollectGit(int notAccessedForDays)
+        {
+            if (!Directory.Exists(AppData.GitRestoreDir))
+            {
+                return;
+            }
+
+            var gitWorkTreeRoots = Directory.EnumerateDirectories(AppData.GitRestoreDir, ".git", SearchOption.AllDirectories);
+
+            using (Progress.Start("GC dependency git repositories"))
```
Comment (yufeih) [inline src/docfx/gc/GarbageCollector.cs]: 'Cleaning git repositories' ?

File: src/docfx/gc/GarbageCollector.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/gc/GarbageCollector.cs]: sure

---
In reply to: [230655577](https://github.com/dotnet/docfx/pull/3571#discussion_r230655577) [](ancestors = 230655577)

File: README.md [hunk 7]
```diff
@@ -40,6 +40,15 @@ docfx build <docset_path> [<options>]
...
   - --github-token: The GitHub token used to get contribution information from GitHub API
 
+### GC
+`gc` command helps you to clean up your local `%DOCFX_APPDATA_PATH%` folder, delete unused downloaded files and git work trees, save your storage usage.
+
+```powershell
+docfx gc [<options>]
+```
+-options:
+  - --retention-days: Only keep files/folder which are accessed/wrote within <retention-days> days.
```
Comment (live1206) [inline README.md]: >wrote [](start = 64, length = 5)

written

File: src/docfx/gc/GarbageCollector.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            var cleaned = 0;
+            if (!Directory.Exists(AppData.UrlRestoreDir))
+            {
+                return cleaned;
+            }
+
+            var downloadedFiles = Directory.EnumerateFiles(AppData.UrlRestoreDir, "*", SearchOption.AllDirectories);
+
+            using (Progress.Start("Cleaning download files"))
```
Comment (live1206) [inline src/docfx/gc/GarbageCollector.cs]: >download [](start = 44, length = 8)

downloaded, naming consistency

**Pull request merged**